src/core/src/core_logic/MaintenanceWindow.py

```diff
@@ -59,9 +59,13 @@ def get_remaining_time_in_minutes(self, current_time=None, log_to_stdout=False):
 
         return remaining_time_in_minutes
 
-    def is_package_install_time_available(self, remaining_time_in_minutes=None):
+    def is_package_install_time_available(self, reboot_manager, remaining_time_in_minutes=None, number_of_packages = 1):
         """Check if time still available for package installation"""
-        cutoff_time_in_minutes = Constants.REBOOT_BUFFER_IN_MINUTES + Constants.PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES
+        cutoff_time_in_minutes = Constants.PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES * number_of_packages
```

Comment:
Current value of Constants.PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES is 5 minutes. This is the expected maximum time taken by installation of a package (along with packages on which it is dependent on). 
For batch size 3, the cutoff_time_in_minutes is 5*3=15 minutes. In this, we are taking into consideration that all the 3 packages (along with packages it is dependent on) might take maximum expected time (5 minutes). I think it is very unlikely scenario that all the three will take maximum expected time. Specially, when we increase the batch size in future, suppose we increase to 6 then the buffer time would be 6*5=30 minutes which is too much as we are considering that all the packages would take maximum expected time.
For lower batch size such as 3, this is not a big issue, but we need to handle it for larger batch size.

Any suggestions here on how should we calculate cutoff time? @kjohn-msft @rane-rajasi @SathishMSFT

Reply:
Since the time taken does not scale linearly for batch installation (the whole point of this PR), you'd need data to understand what function to use instead of straight up multiplication of count against single-package time. See if you can form such a function from the data you already have from experimentation. You'll need to test the function against data you've seen to ensure the margin for error is set up so that the window exceeding is extremely unlikely.

Reply:
@GAURAVRAMRAKHYANI was this looked at further?

Reply:
I am still working on it.

Currently the max expected time for installation of a package is 5 minutes. As per my experiments, most of the packages take only 10-20 seconds to install. I have seen maximum of around 90 seconds for a package to install. I think the 5 minutes cutoff would be very extreme case. For single package it is fine to take max expected time as cutoff time but for multiple packages, taking max expected time for all the packages is very unlikely and assuming that would unnecessarily increase cutoff time.

There are two functions which I can think of:
(a) Take 5 minutes for 1 package and 2 minutes for rest of the packages.
e.g.
if there is 1 package. Then cutoff time would be 5 minutes.
if there are 2 packages. Then cutoff time would be 5+2=7 minutes.
if there are 3 packages. Then cutoff time would be 5+2+2=9 minutes.

(b) Take 5 minutes for half of the packages in the batch.
e.g.
If there is 1 package, the cutoff time is 5 minutes
If there are 2 packages, the cutoff time is 5+2=7 minutes
If there are 3 packages, the cutoff time is 5+2+5=12 minutes
If there are 4 packages, the cutoff time is 5+2+5+2=14 minutes

Reply:
I think option (a) is better because as per testing, the improvement through parallel patching is around 60% if batch size is 3. 

Current maximum time assumed for package installation is 5 minutes. As per improvement of 60%, the maximum time can be assumed to be 40% of 5 = 2 minutes when the packages are installed in batches.

For first package, taking 5 minutes and for rest of the packages, taking 2 minutes would be fine.

PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES is 5
We can introduce new constant PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES_IN_BATCH = 2

cutoff_time_in_minutes  = PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES + (number_of_packages -1)*PACKAGE_INSTALL_EXPECTED_MAX_TIME_IN_MINUTES_IN_BATCH 

number_of_packages =1, cutoff_time_in_minutes  =5
number_of_packages =2, cutoff_time_in_minutes  =5 + 2 = 7
number_of_packages =3, cutoff_time_in_minutes  =5 + 2 +2 = 9

Reply:
@kjohn-msft I discussed with @SathishMSFT on this offline. For first iteration we think it is fine to use multiplication. So, if batch size is 3 then 5*3=15 minutes would be expected maximum time for installation. We think it is better to analyze the telemetry data on maximum time taken to install batch size 3 and then make some formula.

Reply:
@rane-rajasi please capture this thread for review later.